### PR TITLE
ENH: Add ordered subsets MLEM

### PR DESCRIPTION
**Pull request by adler-j**
Adds ordered subsets mlem as a separate option for MLEM, an option would be to merge this with the existing MLEM formulation (basically let users give a list of operators and data, optionally). How do we feel about it?

### CODE REVIEW

File: odl/solvers/iterative/statistical.py [hunk 1]
```diff
@@ -121,6 +138,131 @@ def mlem(op, x, data, niter=1, noise='poisson', callback=None):
         raise RuntimeError('unknown noise model')
 
 
+def osmlem(op, x, data, niter=1, noise='poisson', callback=None, **kwargs):
+    """Ordered Subsets Maximum Likelihood Expectation Maximation algorithm.
+
+    Attempts to solve::
+
+        max_x L(data | x)
```
Comment (kohr-h) [inline odl/solvers/iterative/statistical.py]: Flipped. This is the pure ML problem, but despite its name, the MLEM method tries to maximize the likelihood of the unknown parameters x given the data. See for example [this paper](http://link.springer.com/chapter/10.1007/978-3-319-01712-9_2).

File: odl/solvers/iterative/statistical.py [hunk 2]
```diff
@@ -121,6 +138,131 @@ def mlem(op, x, data, niter=1, noise='poisson', callback=None):
...
+
+    where ``L(data, | x)`` is the likelihood of ``data`` given ``x``. The
+    likelihood depends on the forward operators ``op[0], ..., op[n-1]`` such
+    that (approximately)::
+
+        op[i](x) = data[i]
+
+    With the precise form of *approximately* is determined by ``noise``.
```
Comment (kohr-h) [inline odl/solvers/iterative/statistical.py]: Don't start the sentence with "With"

File: odl/solvers/iterative/statistical.py [hunk 3]
```diff
@@ -121,6 +138,131 @@ def mlem(op, x, data, niter=1, noise='poisson', callback=None):
         raise RuntimeError('unknown noise model')
 
 
+def osmlem(op, x, data, niter=1, noise='poisson', callback=None, **kwargs):
+    """Ordered Subsets Maximum Likelihood Expectation Maximation algorithm.
+
+    Attempts to solve::
```
Comment (kohr-h) [inline odl/solvers/iterative/statistical.py]: Proper sentence?

File: odl/solvers/iterative/statistical.py [hunk 4]
```diff
@@ -121,6 +138,131 @@ def mlem(op, x, data, niter=1, noise='poisson', callback=None):
...
+        \\mathbb{E}(X(A_i(x))) = A_i(x)
+
+    with 'poisson' noise the algorithm is given by partial updates:
+
+    .. math::
+
+       x_{n + m/M} =
+       \\frac{x_{n + (m - 1)/M}}{A_i^* 1} A_i^* (g_i / A_i(x_{n + (m - 1)/M}))
+
+    for :math:`m = 1, ..., M` and :math:`x_{n+1} = x_{n + M/M}`
```
Comment (kohr-h) [inline odl/solvers/iterative/statistical.py]: Final full stop missing.

File: odl/solvers/iterative/statistical.py [hunk 5]
```diff
@@ -121,6 +138,131 @@ def mlem(op, x, data, niter=1, noise='poisson', callback=None):
...
+
+    .. math::
+
+        \prod_{i=1}^M P(g_i | g_i \\text{ is } X(A_i(x)) \\text{ distributed})
+
+    where the expectation of :math:`X(A_i(x))` satisfies
+
+    .. math::
+
+        \\mathbb{E}(X(A_i(x))) = A_i(x)
```
Comment (kohr-h) [inline odl/solvers/iterative/statistical.py]: Could you be a bit more precise here? I don't see where an expectation is maximized here.

File: odl/solvers/iterative/statistical.py [hunk 6]
```diff
@@ -121,6 +138,131 @@ def mlem(op, x, data, niter=1, noise='poisson', callback=None):
...
+        Noise model determining the variant of MLEM.
+        For ``'poisson'``, the initial value of ``x`` should be
+        non-negative.
+    callback : callable, optional
+        Function called with the current iterate after each iteration.
+
+    Other Parameters
+    ----------------
+    sensitivities : float or ``op.domain`` `element-like`, optional
+        Usable with ``noise='poisson'``. The algorithm contains a ``A^T 1``
```
Comment (kohr-h) [inline odl/solvers/iterative/statistical.py]: an ``A^T 1`` term

File: odl/solvers/iterative/statistical.py [hunk 7]
```diff
@@ -121,6 +138,131 @@ def mlem(op, x, data, niter=1, noise='poisson', callback=None):
...
+    loglikelihood : Function for calculating the logarithm of the likelihood
+    """
+    noise, noise_in = str(noise).lower(), noise
+    if noise not in AVAILABLE_MLEM_NOISE:
+        raise NotImplemented("noise '{}' not understood"
+                             ''.format(noise_in))
+
+    n_ops = len(op)
+    if len(data) != n_ops:
+        raise ValueError('number of data does not match number of operators')
```
Comment (kohr-h) [inline odl/solvers/iterative/statistical.py]: print the values

File: odl/solvers/iterative/statistical.py [hunk 8]
```diff
@@ -121,6 +138,131 @@ def mlem(op, x, data, niter=1, noise='poisson', callback=None):
         raise RuntimeError('unknown noise model')
 
 
+def osmlem(op, x, data, niter=1, noise='poisson', callback=None, **kwargs):
```
Comment (kohr-h) [inline odl/solvers/iterative/statistical.py]: Include a citation.

File: odl/solvers/iterative/statistical.py [hunk 9]
```diff
@@ -121,6 +138,131 @@ def mlem(op, x, data, niter=1, noise='poisson', callback=None):
...
+    .. math::
+
+        \\mathbb{E}(X(A_i(x))) = A_i(x)
+
+    with 'poisson' noise the algorithm is given by partial updates:
+
+    .. math::
+
+       x_{n + m/M} =
+       \\frac{x_{n + (m - 1)/M}}{A_i^* 1} A_i^* (g_i / A_i(x_{n + (m - 1)/M}))
```
Comment (kohr-h) [inline odl/solvers/iterative/statistical.py]: Mention that convergence is not guaranteed and there are known pathological cases, but it works very well in practically relevant situations.

File: odl/solvers/iterative/statistical.py [hunk 10]
```diff
@@ -90,18 +97,28 @@ def mlem(op, x, data, niter=1, noise='poisson', callback=None):
     --------
     loglikelihood : Function for calculating the logarithm of the likelihood
     """
+    # TODO: Should this be call to osmlem?
```
Comment (kohr-h) [inline odl/solvers/iterative/statistical.py]: I'd say yes, if we don't create extra temporaries.

File: odl/solvers/iterative/statistical.py [hunk 11]
```diff
@@ -121,6 +138,131 @@ def mlem(op, x, data, niter=1, noise='poisson', callback=None):
...
+                             for opi in op]
+        else:
+            # Make sure the sensitivities is a list of the correct size.
+            try:
+                list(sensitivities)
+            except TypeError:
+                sensitivities = [sensitivities] * n_ops
+
+        tmp_dom = op[0].domain.element()
+        tmp_ran = [opi.range.element() for opi in op]
```
Comment (kohr-h) [inline odl/solvers/iterative/statistical.py]: Can we make these optional? I feel we potentially create a lot of memory overhead here.
How about making `tmps_ran = None` in the non-temp case and have an extra step `tmp_ran = None if tmps_ran is None else tmps_ran[i]` in the loop, and then use that as temporary?

File: odl/solvers/iterative/statistical.py (hunk 5 above)
Comment (kohr-h) [inline odl/solvers/iterative/statistical.py]: Do the same for the regular MLEM solver.

**Comment by kohr-h** [comment]
Oh, and perhaps an example using it would be nice, e.g. extending an existing one for MLEM.

File: odl/solvers/iterative/statistical.py (hunk 8 above)
Comment (adler-j) [inline odl/solvers/iterative/statistical.py]: Dont have one, perhaps someone else does? @aringh @ozanoktem

File: odl/solvers/iterative/statistical.py (hunk 1 above)
Comment (ozanoktem) [inline odl/solvers/iterative/statistical.py]: I would also point to section 5.3.2 in Mathematical Methods in Image Reconstruction, by F. Natterer and F. Wübelling, SIAM, 2001.

File: odl/solvers/iterative/statistical.py (hunk 8 above)
Comment (ozanoktem) [inline odl/solvers/iterative/statistical.py]: Page 122 in [Mathematical Methods in Image Reconstruction, by F. Natterer and F. Wübelling, SIAM, 2001](http://epubs.siam.org/doi/book/10.1137/1.9780898718324) is in my opinion the best reference for OS-MLEM. Short and precise, while pointing out the difficulties in assessing convergence properties.

File: odl/solvers/iterative/statistical.py (hunk 11 above)
Comment (adler-j) [inline odl/solvers/iterative/statistical.py]: This has the same memory use as the regular MLEM algorithm. I'll wait with that until later.

I tried this for the arithmetic operators and it turned out to simply be a lot of work and no one has used it so far.

File: odl/solvers/iterative/statistical.py (hunk 8 above)
Comment (ozanoktem) [inline odl/solvers/iterative/statistical.py]: We may also cite section 6.9.4 in [Medical Image Reconstruction: A Conceptual Tutorial by G. Zeng, Springer-Verlag, 2010](http://www.springer.com/gp/book/9783642053689).

File: odl/solvers/iterative/statistical.py (hunk 11 above)
Comment (adler-j) [inline odl/solvers/iterative/statistical.py]: Frankly the biggest worry is the sensitivities, which has a larger memory use.

**Comment by adler-j** [comment]
Updated by comments, merge?

File: odl/solvers/iterative/statistical.py [hunk 12]
```diff
@@ -36,9 +36,9 @@ def mlem(op, x, data, niter=1, noise='poisson', callback=None, **kwargs):
 
     Attempts to solve::
 
-        max_x L(data | x)
+        max_x L(x | data)
 
-    where ``L(data, | x)`` is the likelihood of ``data`` given ``x``. The
+    where ``L(x | data)`` is the likelihood of ``data`` given ``x``. The
```
Comment (kohr-h) [inline odl/solvers/iterative/statistical.py]: Expression is correct, but the following explanation is still the old one. Should be "the likelihood of ``x`` given ``data``.

File: odl/solvers/iterative/statistical.py [hunk 13]
```diff
@@ -79,79 +79,37 @@ def mlem(op, x, data, niter=1, noise='poisson', callback=None, **kwargs):
...
 
-    where ``L(data, | x)`` is the likelihood of ``data`` given ``x``. The
+    where ``L(x, | data)`` is the likelihood of ``x`` given ``data``. The
     likelihood depends on the forward operators ``op[0], ..., op[n-1]`` such
     that (approximately)::
 
         op[i](x) = data[i]
 
-    With the precise form of *approximately* is determined by ``noise``.
+    Where the precise form of *approximately* is determined by ``noise``.
```
Comment (kohr-h) [inline odl/solvers/iterative/statistical.py]: make "Where" non-capitalized

File: odl/solvers/iterative/statistical.py [hunk 14]
```diff
@@ -187,22 +145,23 @@ def osmlem(op, x, data, niter=1, noise='poisson', callback=None, **kwargs):
...
 
-    for :math:`m = 1, ..., M` and :math:`x_{n+1} = x_{n + M/M}`
+    for :math:`m = 1, ..., M` and :math:`x_{n+1} = x_{n + M/M}`.
+
+    The algorithm is not guaranteed to converge, but works for many practical
+    problems.
+
+    References
+    ----------
+    Natterer, F. Mathematical Methods in Image Reconstruction, section 5.3.2.
```
Comment (kohr-h) [inline odl/solvers/iterative/statistical.py]: I guess you want this to be accessible in an IDE. Still, an entry in the `refs.rst` file would be good. Why not refer to it in the text (e.g. after "problems" just above) and keep this extra reference?

Comment (adler-j) [inline odl/solvers/iterative/statistical.py]: Frankly we need a discussion on refs.rst at some point. I find the fact that I cant find the references while reading the doc quite irritating, and names like [ABC2016] cant be googled either.

What is the gain from having it? A centralized place for references? Avoiding duplication? Is this how others do it in general?

I would personally be in favor of scrapping it.

**Pull request merged**